In [1]:
import pandas as pd 
import numpy as np 
import seaborn as sns 
import matplotlib.pyplot as plt 
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from sklearn import preprocessing
from sklearn import tree 
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn import metrics

In [2]:
data = pd.read_csv("Company_Data.csv")
data.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US  
0         17   Yes  Yes  
1         10   Yes  Yes  
2         12   Yes  Yes  
3         14   Yes  Yes  
4         13   Yes   No

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Sales        400 non-null    float64
 1   CompPrice    400 non-null    int64  
 2   Income       400 non-null    int64  
 3   Advertising  400 non-null    int64  
 4   Population   400 non-null    int64  
 5   Price        400 non-null    int64  
 6   ShelveLoc    400 non-null    object 
 7   Age          400 non-null    int64  
 8   Education    400 non-null    int64  
 9   Urban        400 non-null    object 
 10  US           400 non-null    object 
dtypes: float64(1), int64(7), object(3)
memory usage: 34.5+ KB


In [4]:
data.shape

(400, 11)

In [5]:
data.isnull().sum()

Sales          0
CompPrice      0
Income         0
Advertising    0
Population     0
Price          0
ShelveLoc      0
Age            0
Education      0
Urban          0
US             0
dtype: int64

In [6]:
data.describe()

Sales   CompPrice      Income  Advertising  Population  \
count  400.000000  400.000000  400.000000   400.000000  400.000000   
mean     7.496325  124.975000   68.657500     6.635000  264.840000   
std      2.824115   15.334512   27.986037     6.650364  147.376436   
min      0.000000   77.000000   21.000000     0.000000   10.000000   
25%      5.390000  115.000000   42.750000     0.000000  139.000000   
50%      7.490000  125.000000   69.000000     5.000000  272.000000   
75%      9.320000  135.000000   91.000000    12.000000  398.500000   
max     16.270000  175.000000  120.000000    29.000000  509.000000   

            Price         Age   Education  
count  400.000000  400.000000  400.000000  
mean   115.795000   53.322500   13.900000  
std     23.676664   16.200297    2.620528  
min     24.000000   25.000000   10.000000  
25%    100.000000   39.750000   12.000000  
50%    117.000000   54.500000   14.000000  
75%    131.000000   66.000000   16.000000  
max    191.000000   80.000000   18.000000

In [7]:
type(data)

pandas.core.frame.DataFrame

In [8]:
data.corr()

Sales  CompPrice    Income  Advertising  Population     Price  \
Sales        1.000000   0.064079  0.151951     0.269507    0.050471 -0.444951   
CompPrice    0.064079   1.000000 -0.080653    -0.024199   -0.094707  0.584848   
Income       0.151951  -0.080653  1.000000     0.058995   -0.007877 -0.056698   
Advertising  0.269507  -0.024199  0.058995     1.000000    0.265652  0.044537   
Population   0.050471  -0.094707 -0.007877     0.265652    1.000000 -0.012144   
Price       -0.444951   0.584848 -0.056698     0.044537   -0.012144  1.000000   
Age         -0.231815  -0.100239 -0.004670    -0.004557   -0.042663 -0.102177   
Education   -0.051955   0.025197 -0.056855    -0.033594   -0.106378  0.011747   

                  Age  Education  
Sales       -0.231815  -0.051955  
CompPrice   -0.100239   0.025197  
Income      -0.004670  -0.056855  
Advertising -0.004557  -0.033594  
Population  -0.042663  -0.106378  
Price       -0.102177   0.011747  
Age          1.000000   0.006488  
Education    0.006488   1.000000

In [9]:
data['status'] = data['Sales'].apply(lambda Sales:'Low' if Sales<=7.5 else 'High')

In [10]:
data.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US status  
0         17   Yes  Yes   High  
1         10   Yes  Yes   High  
2         12   Yes  Yes   High  
3         14   Yes  Yes    Low  
4         13   Yes   No    Low

In [11]:
data.drop(['Sales'], axis=1,inplace=True)

In [12]:
data.head()

CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0        138      73           11         276    120       Bad   42   
1        111      48           16         260     83      Good   65   
2        113      35           10         269     80    Medium   59   
3        117     100            4         466     97    Medium   55   
4        141      64            3         340    128       Bad   38   

   Education Urban   US status  
0         17   Yes  Yes   High  
1         10   Yes  Yes   High  
2         12   Yes  Yes   High  
3         14   Yes  Yes    Low  
4         13   Yes   No    Low

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CompPrice    400 non-null    int64 
 1   Income       400 non-null    int64 
 2   Advertising  400 non-null    int64 
 3   Population   400 non-null    int64 
 4   Price        400 non-null    int64 
 5   ShelveLoc    400 non-null    object
 6   Age          400 non-null    int64 
 7   Education    400 non-null    int64 
 8   Urban        400 non-null    object
 9   US           400 non-null    object
 10  status       400 non-null    object
dtypes: int64(7), object(4)
memory usage: 34.5+ KB


In [17]:
label = LabelEncoder()
for i in data.columns:
        print(i)
        if data[i].dtypes ==object:
            data[i] = label.fit_transform(data[i])
        else:
            pass

CompPrice
Income
Advertising
Population
Price
ShelveLoc
Age
Education
Urban
US
status


In [18]:
data.head()

CompPrice  Income  Advertising  Population  Price  ShelveLoc  Age  \
0        138      73           11         276    120          0   42   
1        111      48           16         260     83          1   65   
2        113      35           10         269     80          2   59   
3        117     100            4         466     97          2   55   
4        141      64            3         340    128          0   38   

   Education  Urban  US  status  
0         17      1   1       0  
1         10      1   1       0  
2         12      1   1       0  
3         14      1   1       1  
4         13      1   0       1

In [38]:
x = data.iloc[:,0:10]
y = data.iloc[:,-1]
seed = 7
kfold = KFold(n_splits=8, shuffle=False, random_state=None)

In [39]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3)


In [25]:
#Bagging

In [40]:
cart = DecisionTreeClassifier()
no_trees = 100
model = BaggingClassifier(base_estimator=cart, n_estimators=no_trees, random_state=seed)
model.fit(x_train,y_train)
results = cross_val_score(model, x,y,cv=kfold)
print(results.mean())

0.8075000000000001


In [41]:
preds = model.predict(x_test)
pd.Series(preds).value_counts()

1    62
0    58
dtype: int64

In [42]:
pd.crosstab(y_test,preds)

col_0    0   1
status        
0       48  19
1       10  43

In [43]:
print('Accuracy', np.mean(preds==y_test))

Accuracy 0.7583333333333333


Bagging method accuracy on the test data is 75.83%

In [44]:
# RANDOM FOREST

In [46]:
no_trees = 100
max_features  = 3 
model1 = RandomForestClassifier(n_estimators=no_trees,max_features=max_features)
model1.fit(x_train,y_train)
results1 = cross_val_score(model1,x,y,cv=kfold)
print(results1.mean())

0.8025


In [47]:
preds1 = model1.predict(x_test)
pd.Series(preds1).value_counts()

1    67
0    53
dtype: int64

In [49]:
pd.crosstab(y_test, preds1)

col_0    0   1
status        
0       45  22
1        8  45

In [51]:
print("Accuracy: ", np.mean(preds1==y_test))

Accuracy:  0.75


#Random Forest accuracy on test data is 75%

In [56]:
no_trees1 = 20 
kfold = KFold(n_splits=8, shuffle=False, random_state=None)
model2 = AdaBoostClassifier(n_estimators=no_trees1,random_state=seed)
model2.fit(x_train,y_train)
results2 = cross_val_score(model2, x,y, cv=kfold)
print(results2.mean())

0.8225


In [57]:
preds2 = model2.predict(x_test)
pd.Series(preds2).value_counts()

0    69
1    51
dtype: int64

In [59]:
pd.crosstab(y_test, preds2)

col_0    0   1
status        
0       56  11
1       13  40

In [61]:
print("Accuracy: ", np.mean(preds2==y_test))

Accuracy:  0.8


Adaboost accuracy on test data is 80%

Adaboost is the best model to analyse the Sale of the company
